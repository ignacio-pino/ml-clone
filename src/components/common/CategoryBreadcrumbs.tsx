import { ChevronRightIcon } from "@chakra-ui/icons";
import { Breadcrumb, BreadcrumbItem, BreadcrumbLink } from "@chakra-ui/react";
import { IPathFromRoot } from "../../Interfaces";
import { createSearchParams, useNavigate } from "react-router-dom";

export default function CategoryBreadcrumbs({
  pathFromRoot,
  categoryId,
  includeCurrentCategory,
}: {
  pathFromRoot: IPathFromRoot[];
  categoryId?: string;
  includeCurrentCategory?: boolean;
}) {
  const navigate = useNavigate();

  const handleClick = (event: React.MouseEvent) => {
    event?.preventDefault();
    const params = { category: event.currentTarget.id };
    navigate({
      pathname: "/products",
      search: `${createSearchParams(params)}`,
    });
  };

  return (
    <Breadcrumb
      spacing="8px"
      separator={<ChevronRightIcon color="gray.500" />}
      fontSize={"smaller"}
      maxW={"70%"}
    >
      {pathFromRoot?.map(({ id, name }) => {
        return id === categoryId && !includeCurrentCategory ? null : (
          <BreadcrumbItem>
            <BreadcrumbLink id={id} onClick={handleClick}>
              {" "}
              {name}
            </BreadcrumbLink>
          </BreadcrumbItem>
        );
      })}
    </Breadcrumb>
  );
}
