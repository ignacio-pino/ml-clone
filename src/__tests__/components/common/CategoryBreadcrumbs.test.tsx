import { render, screen, fireEvent } from "@testing-library/react";
import CategoryBreadcrumbs from "../../../components/common/CategoryBreadcrumbs";
import { MemoryRouter, createSearchParams } from "react-router-dom";

const mockedUseNavigate = vitest.fn();

vitest.mock("react-router-dom", () => ({
  ...require("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
}));

describe("CategoryBreadcrumbs Component", () => {
  const mockPathFromRoot = [
    { id: "1", name: "Root" },
    { id: "2", name: "Level1" },
    { id: "3", name: "Level2" },
  ];

  beforeEach(() => {
    mockedUseNavigate.mockClear();
  });

  it("matches snapshot", () => {
    const params = { category: "test" };
    const container = render(
      <MemoryRouter
        initialEntries={[
          { pathname: "/products", search: `${createSearchParams(params)}` },
        ]}
      >
        <CategoryBreadcrumbs pathFromRoot={mockPathFromRoot} categoryId="3" />
      </MemoryRouter>
    ).container;

    expect(container).toMatchSnapshot();
  });

  it("renders the breadcrumbs correctly", () => {
    render(
      <MemoryRouter>
        <CategoryBreadcrumbs pathFromRoot={mockPathFromRoot} categoryId="3" />
      </MemoryRouter>
    );

    const rootBreadcrumb = screen.getByText("Root");
    const level1Breadcrumb = screen.getByText("Level1");
    const level2Breadcrumb = screen.queryByText("Level2");

    expect(rootBreadcrumb).toBeInTheDocument();
    expect(level1Breadcrumb).toBeInTheDocument();
    expect(level2Breadcrumb).toBeNull();
  });

  it("calls the handleClick function with correct params when breadcrumb is clicked", () => {
    render(
      <MemoryRouter>
        <CategoryBreadcrumbs pathFromRoot={mockPathFromRoot} categoryId="3" />
      </MemoryRouter>
    );

    const rootBreadcrumb = screen.getByText("Root");

    fireEvent.click(rootBreadcrumb);

    expect(mockedUseNavigate).toHaveBeenCalledWith({
      pathname: "/products",
      search: "category=1",
    });
  });

  it("doesn't render the breadcrumb for the current category", () => {
    render(
      <MemoryRouter>
        <CategoryBreadcrumbs pathFromRoot={mockPathFromRoot} categoryId="2" />
      </MemoryRouter>
    );

    const level2Breadcrumb = screen.getByText("Level2");
    const level1Breadcrumb = screen.queryByText("Level1");

    expect(level2Breadcrumb).toBeInTheDocument();
    expect(level1Breadcrumb).toBeNull();
  });
});
