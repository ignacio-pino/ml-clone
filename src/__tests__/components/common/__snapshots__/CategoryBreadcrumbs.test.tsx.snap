// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CategoryBreadcrumbs Component > matches snapshot 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="chakra-breadcrumb css-1rsh7lu"
  >
    <ol
      class="chakra-breadcrumb__list css-70qvj9"
    >
      <li
        class="chakra-breadcrumb__list-item css-18biwo"
      >
        <a
          class="chakra-breadcrumb__link css-0"
          id="1"
        >
           
          Root
        </a>
        <span
          class="css-syhbcg"
          role="presentation"
        >
          <svg
            class="chakra-icon css-2u1jce"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              fill="currentColor"
            />
          </svg>
        </span>
      </li>
      <li
        class="chakra-breadcrumb__list-item css-18biwo"
      >
        <a
          class="chakra-breadcrumb__link css-0"
          id="2"
        >
           
          Level1
        </a>
      </li>
    </ol>
  </nav>
</div>
`;
